tensorboard/webapp/metrics/views/card_renderer/scalar_card_container.ts

```diff
@@ -222,6 +265,106 @@ export class ScalarCardContainer implements CardRenderer, OnInit {
       distinctUntilChanged(areSeriesDataListEqual)
     );
 
+    function getSmoothedSeriesId(seriesId: string): string {
+      return JSON.stringify(['smoothed', seriesId]);
+    }
+
+    this.dataSeries$ = runIdAndPoints$.pipe(
+      combineLatestWith(this.store.select(getMetricsScalarSmoothing)),
+      switchMap(([runsData, smoothing]) => {
+        const dataSeriesList = runsData.map(({runId, points}) => {
+          return {id: runId, points};
+        });
+        if (smoothing === 0) {
+          return of(dataSeriesList);
+        }
+        return from(classicSmoothing(dataSeriesList, smoothing)).pipe(
+          map((smoothedDataSeriesList) => {
+            const dataSeriesMap = new Map(
+              dataSeriesList.map((datum) => [datum.id, datum])
+            );
+            for (const smoothedSeries of smoothedDataSeriesList) {
+              const {points: origPoints} = dataSeriesMap.get(
+                smoothedSeries.id
+              )!;
+              dataSeriesList.push({
+                id: getSmoothedSeriesId(smoothedSeries.id),
+                points: smoothedSeries.points.map(({y}, index) => {
+                  return {...origPoints[index], y};
+                }),
+              });
+            }
+            return dataSeriesList;
+          })
+        );
+      }),
+      distinctUntilChanged(areSeriesEqual),
+      startWith([])
+    );
+
+    this.chartMetadataMap$ = runIdAndPoints$.pipe(
+      switchMap((runIdAndPoints) => {
+        if (!runIdAndPoints.length) {
+          return of([]);
+        }
+
+        return combineLatest(
+          runIdAndPoints.map((runIdAndPoint) => {
+            return this.getRunDisplayNameAndPoints(runIdAndPoint);
+          })
+        );
+      }),
+      combineLatestWith(
+        this.store.select(getCurrentRouteRunSelection),
+        this.store.select(getRunColorMap),
+        this.store.select(getMetricsScalarSmoothing)
+      ),
+      // When the `fetchRunsSucceeded` action fires, the run selection
+      // map and the metadata change. To prevent quick fire of changes,
+      // debounce by a microtask to emit only single change for the runs
+      // store change.
+      debounceTime(0),
```

Comment:
Three things:

- I had originally had the code that way but it was reading very poorly
- `e.g. might it be possible for the line chart to render lots of dataSeries$ in an initial pass with empty metadata, and then re-render it again 1 tick later?` Converse is more likely as we have `async` function for smoothing.
- Because line chart expects two different props for data vs. metadata, it is only obvious that they can diverge and we need to guard against it (and in fact, we do guard against it). 

For readability and for practical reasons, I will not make the change unless it proves to be a problem (in terms of UX or performance).